cli/command_repository_upgrade.go

```diff
@@ -72,9 +86,123 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
 
 	rollbackCmd.Action(svc.directRepositoryWriteAction(c.forceRollbackAction))
 
+	validateCmd := parent.Command("validate", "Validate the upgraded indexes.")
+
+	validateCmd.Action(svc.directRepositoryWriteAction(c.validateAction))
+
 	c.svc = svc
 }
 
+// assign store the info struct in a map that can be used to compare indexes
+func assign(iif content.Info, i int, m map[content.ID][2]index.Info) {
+	v := m[iif.GetContentID()]
+	v[i] = iif
+	m[iif.GetContentID()] = v
+}
+
+// loadIndexBlobs load index blobs into indexEntries map
+func loadIndexBlobs(ctx context.Context, indexEntries map[content.ID][2]index.Info, sm *content.SharedManager, index int, indexBlobInfos []content.IndexBlobInfo) error {
+	d := gather.WriteBuffer{}
+	for _, indexBlobInfo := range indexBlobInfos {
+		blobID := indexBlobInfo.BlobID
+		indexInfos, err := sm.LoadIndexBlob(ctx, blobID, d)
+		if err != nil {
+			return errors.Wrapf(err, "failed to load index blob with BlobID %s", blobID)
+		}
+		for _, indexInfo := range indexInfos {
+			assign(indexInfo, index, indexEntries)
+		}
+	}
+	return nil
+}
+
+// validateAction returns an array of strings (report) that describes differences between
+// the V0 index blob and V1 index blob content.  This is used to check that the upgraded index
+// (V1 index) reflects the content of the old V0 index.
+func (c *commandRepositoryUpgrade) validateAction(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+
+	indexEntries := map[content.ID][2]index.Info{}
+
+	sm := rep.ContentManager().SharedManager
+
+	indexBlobInfos, _, err := sm.IndexReaderV0().ListIndexBlobInfos(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list index blobs for v0 index")
+	}
+
+	err = loadIndexBlobs(ctx, indexEntries, sm, 0, indexBlobInfos)
+	if err != nil {
+		return errors.Wrapf(err, "failed to load index entries for v0 index entry")
+	}
+
+	indexBlobInfos, _, err = sm.IndexReaderV1().ListIndexBlobInfos(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list index blobs for v1 index")
+	}
```

Comment:
Hmm don't the two blob reader versions need to be inputs to this function? Say, assuming some hypothetical future upgrade from version X to version Y, we still want to verify the before-upgrade indices are equivalent to the after-upgrade indices?

Reply:
@redgoat650 in theory yes, but that would be designing before we have a use-case.  For now just dealing with v0 -> v1 index upgrade.